### DISCUSSION TITLE
Over riding classes - substitute in Bootstrap?

### DISCUSSION THREAD

**Discussion by goclimbing** [woocommerce/woocommerce#52913] [Checkout Flow]
For instance, on the Proceed to Checkout in the default cart template I see no way to over ride css to use bootstrap button classes.  If I add them they are appended and no styling is changed.

So what's the best approach to adding custom CSS or changing class/ids in these templates?

**Comment by nielslange** [comment]
Hello @goclimbing 👋

Thanks for reaching out. I briefly looked at this issue and was able to reproduce your observation. When applying the Bootstrap styles to the _"Proceed to Checkout"_ button, the default WooCommerce styles will not be overwritten. Apparently, the CSS specificity of the WooCommerce styles is higher than the one of the Bootstrap styles. To overwrite the WooCommerce styles, you would have to remove the corresponding CSS classes from the button. This could be accomplished using a JS snippet. Something like this:
```js
// Select the element based on one of the existing classes
const checkoutButton = document.querySelector('.wc-block-components-button');

// Remove the specified class
checkoutButton.classList.remove( 'wp-element-button' );

// Add the desired classes
checkoutButton.classList.add('btn', 'btn-primary');
```

Removing the class `wp-element-button` and adding the classes ` btn btn-primary` lead to this result:

<img width="1512" alt="Screenshot 2024-11-18 at 23 29 25" src="https://github.com/user-attachments/assets/fb54e525-cba8-49cf-bda6-6068bd573c7b">

That said, if you're using a block theme, you can also adjust the style of the button directly via `WP Admin > Appearance > Editor > Pages > Cart`. Then, click on the global styles icon (two icons left of the save button in the upper-right corner). Next, select `Colors > Button` to style the button according to your needs. This is how the corresponding pages looks like:

<img width="1512" alt="Screenshot 2024-11-18 at 23 40 20" src="https://github.com/user-attachments/assets/319e1905-02e7-40e6-8cbd-99b15013a8d2">

I hope this helps to style the button according to your needs.

**Reply by goclimbing** [reply]
HI @nielslange  - thank you for the reply. I tried the javascript and tho everything looks correct I'm getting a JS error.

<img width="880" alt="Screenshot 2024-11-18 at 12 32 10 PM" src="https://github.com/user-attachments/assets/1df2f806-4a42-43c3-870c-904eadca17a5">

<img width="646" alt="Screenshot 2024-11-18 at 12 31 37 PM" src="https://github.com/user-attachments/assets/5602c3bb-794d-4c0b-81aa-63975c7fcd4b">
That's from the second line where it Removes the class.

For what it's worth I'm working on a full custom Woo build using Understrap. I thought about taking the javascript route to apply needed bootstrap styles but ended up applying some styles with hooks.  Then I got to working the cart and checkout page only to learn it's template based. I guess it's been that long since I've worked with a new Woo build.

I was rather surprised to find no modificaitons to the cart and checkout code did anything :) 

I'm considering just using the classic cart and checkout as I feel like it'll give me the control I need. For instance, I need to be able to update the price using AJAX/Jquery for items in the cart and at checkout.  We have a plugin that pulls in updated pricing (the site is for metals, bullion and coins).

**Reply by nielslange** [reply]
> I tried the javascript and tho everything looks correct I'm getting a JS error.

@goclimbing I've looked up the script, and it loads a bit too early. The following script ensures that it loads once the page is loaded:

```js
window.onload = function () {
  // Select the element based on one of the existing classes
  const checkoutButton = document.querySelector('.wc-block-components-button');

  if ( checkoutButton ) {
    // Remove the specified class
    checkoutButton.classList.remove('wp-element-button');

    // Add the desired classes
    checkoutButton.classList.add('btn', 'btn-primary');	
  }
}
```

> I was rather surprised to find no modificaitons to the cart and checkout code did anything :)

Do you mean in terms of CSS modifications? If that's the case, you could either overwrite the CSS styles according to your needs or remove the unwanted and add the wanted CSS classes using the JS snippet I shared above. 

> I'm considering just using the classic cart and checkout as I feel like it'll give me the control I need. For instance, I need to be able to update the price using AJAX/Jquery for items in the cart and at checkout. We have a plugin that pulls in updated pricing (the site is for metals, bullion and coins).

Can you share more details regarding this use case? And potentially share the code that you are using to accomplish this? Once I know your specific use case, I might be able to provide a solution that works with the Cart and the Checkout blocks.

**Reply by goclimbing** [reply]
@neilmccreadie  thanks for that update on the js. it works.

> Do you mean in terms of CSS modifications? If that's the case, you could either overwrite the CSS styles according to your needs or remove the unwanted and add the wanted CSS classes using the JS snippet I shared above.

I mean the php code in templates/cart and templates/checkout - I was tweaking cart.php and a few other files and no output changed. that's when i realized those two areas are these templates.

Over writing the css classes isn't the direction i'm going. i'm leaning hard into bootstrap - have a bunch of ACF/Gutenberg blocks all wrapped nicely in Bootstrap, etc. 

The add to cart button in both the single and loop needed updating to bootstrap. The `[child theme dir]/inc/woocommerce.php` support file that comes with Understrap includes code like:

```php
if ( ! function_exists( 'understrap_loop_add_to_cart_args' ) ) {

    /**
     * Add Bootstrap button classes to add to cart link.
     *
     * @param array<string,mixed> $args Array of add to cart link arguments.
     * @return array<string,mixed> Array of add to cart link arguments.
     */
    function understrap_loop_add_to_cart_args( $args ) {
        if ( isset( $args['class'] ) && ! empty( $args['class'] ) ) {
            if ( ! is_string( $args['class'] ) ) {
                return $args;
            }

            // Remove the `button` class if it exists.
            if ( false !== strpos( $args['class'], 'button' ) ) {
                $args['class'] = explode( ' ', $args['class'] );
                $args['class'] = array_diff( $args['class'], array( 'button' ) );
                $args['class'] = implode( ' ', $args['class'] );
            }

            $args['class'] .= ' btn btn-secondary text-white';
        } else {
            $args['class'] = 'btn btn-outline-primary btn-secondary';
        }

        if ( 'bootstrap4' === get_theme_mod( 'understrap_bootstrap_version', 'bootstrap4' ) ) {
            $args['class'] .= ' btn-block';
        }

        return $args;
    }

}
```


(weird, can't get it to format as code correctly).  

So no tweaking of the add to cart button php needed.

At the end of the day I'll use the JS (thank you again) as I have some plugins like AvaTax I won't be updating to Bootstrap.  

And we may stick with the templates but I have to get with my IT team to discuss.

For what it's worth it's a site selling metals, bullion and coins.  We have a third party API provider that pulls in products from our metals vendor to populate the woocommerce product database.  Only product title, product images, initial price, and internal to vendor sku's etc, are pulled in.  

The price is updated via the API every two minutes.  So I need wrappers around anywhere there is a price so my AJAX/JQuery script can update the price in the shop/cart/checkout areas (they will blink like a stock ticker red or green).

No descriptions were included so we used ChatGPT to populate the SEO, short and long descriptions.

New products might appear but w/o AvaTax tax code. So I'm checking woo products db for products missing tax codes then calling and API endpoint directly to grab the new code and populate.

**Reply by nielslange** [reply]
Thanks for providing such a detailed explanation of your use case and the need for prices to be updated dynamically using AJAX/jQuery. This functionality is not currently available in WooCommerce out of the box. I recommend creating an issue on https://github.com/woocommerce/woocommerce/issues, where you can describe the expected outcome and, if possible, share an example implementation or a more detailed use case.
